### PR TITLE
Implement classmethod pytype

### LINKED ISSUE #1693: Support classmethod
I have a workaround to create classmethods in PyBind.
```
  template<class Func, typename ... Args>
  pybind11::object classmethod(Func f, Args ... args) {
    pybind11::object cf = pybind11::cpp_function(f, args...);
    return pybind11::object(PyClassMethod_New(cf.ptr()), true);
  }
```
used like this
```
void foo(py::object cls, ...);
cls.attr("foo") = classmethod(&foo);
```
Be useful to have such mechanism in pybind proper

### PR DESCRIPTION
**Pull request by Skylion007**
## Description
I've been mulling over how to implement this for a while now and finally got it working without segfaults.

Closes #1693 and implements classmethod support in pybind11. This allows defining classmethods which have the class of the caller automatically prepended to the list of args. This is nice because it allows adding/modifying class variables through the python bindings directly without a helper C++ class or struct.

NB: One thing I have noticed during my testing is that while `__new__` has a classmethod signature, it is actually implemented as a static method. Therefore trying to add a classmethod onto `__new__` will cause errors.

@rwgk I would like some thoughts on how to think about designing some proper tests for this functionality as there are probably a lot of edge cases having to do with potential mem leaks and other odd behavior. We also do play fast and loose with the terminology of classmethod with our add_classmethod function in pybind11 so I might need to refactor this. 

Tests that may need to be added:
1. memleaks? (I am not if the added attributes are properly cleaned up).
2. docstring
3. overloading precedence / siblings

Additional things needed: documentation.

## Suggested changelog entry:

```rst
* Implement classmethod support
```

### CODE REVIEW

**Comment by Skylion007** [comment]
Seems pypy doesn't expose the PyClassMethod_Type, is there a stable API way to implement the check_ method? Or even a good way to test it to verify it works. @rwgk @henryiii thoughts ?

**Comment by rwgk** [comment]
> Seems pypy doesn't expose the PyClassMethod_Type, is there a stable API way to implement the check_ method? Or even a good way to test it to verify it works. @rwgk @henryiii thoughts ?

@mattip for the PyPy PyClassMethod_Type question

(I can look at this PR only later.)

**Comment by Skylion007** [comment]
Interestingly noticed it's not part of the CPython Stable-API, which is really odd for a pytype. It appears that CPython wants projects to move to the PyMethodDef API, but that only supports 3.9 and higher: https://docs.python.org/3/c-api/structures.html#c.PyMethodDef

File: tests/test_class.cpp [hunk 1]
```diff
@@ -62,7 +62,15 @@ TEST_SUBMODULE(class_, m) {
     };
 
     py::class_<NoConstructor>(m, "NoConstructor")
-        .def_static("new_instance", &NoConstructor::new_instance, "Return an instance");
+        .def_static("new_instance", &NoConstructor::new_instance, "Return an instance")
+        .def_classmethod(
+            "new_instance_uuid",
```
Comment (rwgk) [inline tests/test_class.cpp:67]: Maybe `serial` or `seq` or similar? (It doesn't look like a universally unique identifier.)

File: tests/test_class.py [hunk 2]
```diff
@@ -31,6 +31,12 @@ def test_instance_new(msg):
     assert cstats.alive() == 0
 
 
+def test_classmethod(num_instances=10):
+    for i in range(num_instances):
+        assert getattr(m.NoConstructor, "uuid", 0) == i
+        m.NoConstructor.new_instance_uuid()
+
```
Comment (rwgk) [inline tests/test_class.py:38]: Maybe (untested):
```
    assert no hasattr(m.NoConstructor, "uuid")
    for i in range(num_instances):
        m.NoConstructor.new_instance_uuid()
        assert m.NoConstructor.uuid == i + 1
```

**Review by rwgk — COMMENTED** [review]
> Tests that may need to be added:
> memleaks? (I am not if the added attributes are properly cleaned up).

GHA valgrind is happy, ASAN & MSAN in the Google environment, too. That's a great start.

The only other thing I'd do, because it's super easy and conclusive beyond a doubt: locally modify the new test(s) by inserting `while True:`, run the test, in another shell (on the same machine) run the top command for a couple minutes and watch `RES` to see if it is stable for the test.

> overloading precedence / siblings

I'd add a test similar to what you have, but with one argument, as a way to ensure that `extra...` is plumbed through.

TBH I need to learn what siblings are.

I wouldn't worry about overload precedence, that logic is somewhere completely different and I don't see how the new `.def_classmethod` could break anything there.

File: include/pybind11/pytypes.h [hunk 3]
```diff
@@ -2089,6 +2090,11 @@ class staticmethod : public object {
     PYBIND11_OBJECT_CVT(staticmethod, object, detail::PyStaticMethod_Check, PyStaticMethod_New)
 };
 
+class classmethod : public object {
```
[the remark this answers is not in the document]
Comment (Skylion007) [reply include/pybind11/pytypes.h:2093]: Yes, it is used in def_classmethod ?

[the remark this answers is not in the document]
Comment (rwgk) [reply include/pybind11/pytypes.h:2093]: Oh, sorry, now I see it. (I was staring at the cpp_function there too much and overlooked the much simpler line below.)

**Comment by virtuald** [comment]
Seems useful. Just add some docs and skip the old versions of pypy and we're done here?

File: tests/test_class.py (hunk 2 above)
Comment (rwgk) [reply tests/test_class.py:39]: Oh ...
Commit 2e3d29d from two months ago escaped my attention.
If you reply with Done or similar here it's more likely that that doesn't happen.

**Comment by rwgk** [comment]
> Seems useful. Just add some docs and skip the old versions of pypy and we're done here?

Agreed.

**Comment by virtuald** [comment]
FWIW, using this locally, docstring has 'self' as first parameter instead of 'cls'. The type is correct however. pybind11-stubgen will also need a fix to generate the classmethod correctly.

**Comment by virtuald** [comment]
Here's a format-patch for the docstring:

```patch
From 7ebb255157d077440956fe9721bdd33ca364c1d1 Mon Sep 17 00:00:00 2001
From: Dustin Spicuzza <dustin@virtualroadside.com>
Date: Sat, 24 Dec 2022 14:39:47 -0500
Subject: [PATCH] Fix classmethod docstring

---
 include/pybind11/attr.h     | 16 ++++++++++++++++
 include/pybind11/pybind11.h |  2 +-
 tests/test_class.py         |  2 ++
 3 files changed, 19 insertions(+), 1 deletion(-)

diff --git a/include/pybind11/attr.h b/include/pybind11/attr.h
index db7cd8ef..53231e65 100644
--- a/include/pybind11/attr.h
+++ b/include/pybind11/attr.h
@@ -26,6 +26,12 @@ struct is_method {
     explicit is_method(const handle &c) : class_(c) {}
 };
 
+/// Annotation for classmethods
+struct is_classmethod {
+    handle class_;
+    explicit is_classmethod(const handle &c) : class_(c) {}
+};
+
 /// Annotation for operators
 struct is_operator {};
 
@@ -426,6 +432,16 @@ struct process_attribute<is_method> : process_attribute_default<is_method> {
     }
 };
 
+/// Process an attribute which indicates that this function is a classmethod
+template <>
+struct process_attribute<is_classmethod> : process_attribute_default<is_classmethod> {
+    static void init(const is_classmethod &s, function_record *r) {
+        r->is_method = true;
+        r->scope = s.class_;
+        r->args.emplace_back("cls", nullptr, handle(), /*convert=*/true, /*none=*/false);
+    }
+};
+
 /// Process an attribute which indicates the parent scope of a method
 template <>
 struct process_attribute<scope> : process_attribute_default<scope> {
diff --git a/include/pybind11/pybind11.h b/include/pybind11/pybind11.h
index 9fd4e214..bd2fd548 100644
--- a/include/pybind11/pybind11.h
+++ b/include/pybind11/pybind11.h
@@ -1593,7 +1593,7 @@ public:
     class_ &def_classmethod(const char *name_, Func &&f, const Extra &...extra) {
         cpp_function cf(std::forward<Func>(f),
                         name(name_),
-                        is_method(*this),
+                        is_classmethod(*this),
                         sibling(getattr(*this, name_, none())),
                         extra...);
         auto cf_name = cf.name();
diff --git a/tests/test_class.py b/tests/test_class.py
index dca01b9d..31fb6b33 100644
--- a/tests/test_class.py
+++ b/tests/test_class.py
@@ -37,6 +37,8 @@ def test_classmethod(num_instances=10):
         m.NoConstructor.new_instance_seq_id()
         assert m.NoConstructor.seq_id == i + 1
 
+    assert m.NoConstructor.new_instance_seq_id.__doc__.startswith("new_instance_seq_id(cls: type) ->")
+
 
 def test_type():
     assert m.check_type(1) == m.DerivedClass1
-- 
2.36.1
```

**Comment by virtuald** [comment]
Just kidding, that doesn't work if you specify arguments. Updated comment. It seems like the two options are to modify the function record (triggers an internals API bump?) or go this route. This seems less invasive, but it's also less consistent with how args are done elsewhere.
